prac_09/taxi_simulator.py

```diff
@@ -0,0 +1,55 @@
+"""
+CP1404/CP5632 Practical - taxi simulator.
+"""
+
+from prac_09.taxi import Taxi
+from prac_09.silver_service_taxi import SilverServiceTaxi
+
+MENU = "q)uit, c)hoose taxi, d)rive"
+
+
+def main():
+    """Program to select taxis and drive, calculating the cost of the journey."""
+    total_bill = 0
+    taxis = [Taxi('Prius', 100), SilverServiceTaxi("Limo", 100, 2), SilverServiceTaxi("Hummer" , 200, 4)]
+    current_taxi = None
+    print("Let's drive!")
+    print(MENU)
+    choice = input("Choose a menu option: ").lower()
+    while choice != "q":
+        if choice == "c":
+            print("Taxis available")
+            display_taxi(taxis)
+            taxi_choice = int(input("Choose a taxi: "))
+            try:
+                current_taxi = taxis[taxi_choice]
+            except IndexError:
+                print("Invalid taxi")
+
+        elif choice == "d":
+            if current_taxi:
+                current_taxi.start_fare()
+                distance = float(input("Drive how far: "))
+                current_taxi.drive(distance)
+                trip_cost = current_taxi.get_fare()
+                print(f"Your {current_taxi.name} trip cost you ${trip_cost:.2f}")
+                total_bill += trip_cost
+            else:
+                print("You need to choose a taxi before you can drive")
+        else:
+            print("invalid option.")
+        print(f"Bill to date: ${total_bill:.2f}")
+        print(MENU)
+        choice = input("Choose a menu option: ").lower()
+    print(f"Total trip cost: ${total_bill:.2f}")
+    print("Taxis are now:")
+    display_taxi(taxis)
+
+
+def display_taxi(taxis):
+    """Display list of taxis."""
+    for i, taxi in enumerate(taxis):
+        print(f"{i} - {taxi}")
+
+
```

Comment:
All your variable names follow the correct naming conventions and are descriptive, also the structure looks good as well. One suggestion for improvement would be to make the error handling more robust. Currently, if the user inputs an invalid taxi choice, the program prints an error message but doesn't reset current_taxi to None in the event of an invalid choice. This could potentially lead to unexpected behavior if the user tries to drive with an invalid taxi selected.
To improve this, consider adding current_taxi = None inside the except block to ensure that the user cannot proceed with an invalid taxi choice. But very well executed overall :)

Reply:
As of right now the code will crash if you put in something different from an int which I could fix, but current_taxi is set to none in my code so if someone tries to set an invalid interger as the taxi the code doesnt assign it to the varaiable and it will stay none, so adding current_taxi = none to the except block is unescessary

appreciate the feedback though...

Reply:
Sorry Scott for the bad feedback in places, I didn't see they were equivalent phrases my bad! Anyway, really good prac and thanks for letting me know.

Reply:
Its all good, it's not bad feedback just a misundestanding. I still appreciate your feedback at least you answer my mention 👍

Reply:
Its all good, it's not bad feedback just a misundestanding. I still appreciate your feedback at least you answer my mention 👍